python/cudf/cudf/core/dataframe.py

```diff
@@ -6323,6 +6323,146 @@ def corr(self):
+        Append rows of `other` to the end of caller, returning a new object.
+        Columns in `other` that are not in the caller are added as new columns.
+
+        Parameters
+        ----------
+        other : DataFrame or Series/dict-like object, or list of these
+            The data to append.
+        ignore_index : bool, default False
+            If True, do not use the index labels.
+        sort : bool, default False
+            Sort columns ordering if the columns of
+            `self` and `other` are not aligned.
+        verify_integrity : bool, default False
+            This Parameter is currently not supported.
+
+        Returns
+        -------
+        DataFrame
+
+        See Also
+        --------
+        cudf.concat : General function to concatenate DataFrame or
+            objects.
+
+        Notes
+        -----
+        If a list of dict/series is passed and the keys are all contained in
+        the DataFrame's index, the order of the columns in the resulting
+        DataFrame will be unchanged.
+        Iteratively appending rows to a cudf DataFrame can be more
+        computationally intensive than a single concatenate. A better
+        solution is to append those rows to a list and then concatenate
+        the list with the original DataFrame all at once.
+        `verify_integrity` parameter is not supported yet.
+
+        Examples
+        --------
+        >>> import cudf
+        >>> df = cudf.DataFrame([[1, 2], [3, 4]], columns=list('AB'))
+        >>> df
+           A  B
+        0  1  2
+        1  3  4
+        >>> df2 = cudf.DataFrame([[5, 6], [7, 8]], columns=list('AB'))
+        >>> df2
+           A  B
+        0  5  6
+        1  7  8
+        >>> df.append(df2)
+           A  B
+        0  1  2
+        1  3  4
+        0  5  6
+        1  7  8
+
+        With `ignore_index` set to True:
+        >>> df.append(df2, ignore_index=True)
+           A  B
+        0  1  2
+        1  3  4
+        2  5  6
+        3  7  8
+
+        The following, while not recommended methods for generating DataFrames,
+        show two ways to generate a DataFrame from multiple data sources.
+        Less efficient:
+        >>> df = cudf.DataFrame(columns=['A'])
+        >>> for i in range(5):
+        ...     df = df.append({'A': i}, ignore_index=True)
+        >>> df
+           A
+        0  0
+        1  1
+        2  2
+        3  3
+        4  4
+
+        More efficient than above:
+        >>> cudf.concat([cudf.DataFrame([i], columns=['A']) for i in range(5)],
+        ...           ignore_index=True)
+           A
+        0  0
+        1  1
+        2  2
+        3  3
+        4  4
+        """
+        if verify_integrity not in (None, False):
+            raise NotImplementedError(
+                "verify_integrity parameter is not supported yet."
+            )
+
+        if isinstance(other, (cudf.Series, dict)):
+            if isinstance(other, dict):
+                other = Series(other)
```

Comment:
This doesn't seem safe. `other` could be a dict of mixed dtypes, i.e. `{'a': 1, 'b': 2.0, 'c': 'blah'}` that maps to the three columns of different dtypes.

Reply:
Yes, `Series` initialization using `dict` is a recent addition to cudf, and the above scenario will fail with an error at pyarrow:

```python
>>> import cudf
>>> s = cudf.Series({'a': 1, 'b': 2.0, 'c': 'blah'})
Traceback (most recent call last):
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.7/site-packages/cudf/core/column/column.py", line 1595, in as_column
    memoryview(arbitrary), dtype=dtype, nan_as_null=nan_as_null
TypeError: memoryview: a bytes-like object is required, not 'dict_values'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.7/site-packages/cudf/core/column/column.py", line 1617, in as_column
    else nan_as_null,
  File "pyarrow/array.pxi", line 269, in pyarrow.lib.array
  File "pyarrow/array.pxi", line 38, in pyarrow.lib._sequence_to_array
  File "pyarrow/error.pxi", line 85, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Could not convert blah with type str: tried to convert to double

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.7/site-packages/cudf/core/series.py", line 173, in __init__
    data.values(), nan_as_null=nan_as_null, dtype=dtype
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.7/site-packages/cudf/core/column/column.py", line 1636, in as_column
    nan_as_null=nan_as_null,
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.7/site-packages/cudf/core/column/column.py", line 1550, in as_column
    pa.Array.from_pandas(arbitrary), dtype=arbitrary.dtype
  File "pyarrow/array.pxi", line 805, in pyarrow.lib.Array.from_pandas
  File "pyarrow/array.pxi", line 265, in pyarrow.lib.array
  File "pyarrow/array.pxi", line 80, in pyarrow.lib._ndarray_to_array
  File "pyarrow/error.pxi", line 85, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: only handle 1-dimensional arrays

>>> import pandas as pd
>>> pd.Series({'a': [1], 'b': [2.0], 'c': ['blah']})
a       [1]
b     [2.0]
c    [blah]
dtype: object

>>> pd.Series({'a': 1, 'b': 2.0, 'c': 'blah'})
a       1
b       2
c    blah
dtype: object
```

I think the main use-case of `Series` initialization using `dict` is having nested lists in a column, so do you think we should add validation to the values in dict before we call `as_column`?

Reply:
I don't think that does what you think it does. Where pdf is a basic one row dataframe, and pds is similar to the Series you defined where each element is a list:
```
In [122]: pdf.append(pds, ignore_index=True)
Out[122]:
     a      b        c
0    1      2     blah
1  [2]  [3.0]  [blah2]
```

Reply:
And similarly with a dict:
```
In [124]: pdf.append({'a': 2, 'b': 3.0, 'c': 'blah2'}, ignore_index=True)
Out[124]:
   a    b      c
0  1  2.0   blah
1  2  3.0  blah2
```

Reply:
Given that the dtypes of the dict need to be potentially homogenous in order to append properly, we should convert to a DataFrame instead of a Series, and then use concatenate against that DataFrame.